Added home page signup popup

aliceriot opened the pull request:
#### What are the relevant tickets?

closes #917 

#### What's this PR do?

This adds a little dialog to the home page and to the signup pages which lets you pick a program. We persist this to `localStorage`. In the dashboard, when the `App` container updates it checks for this value in `localStorage` and, if present, it uses it to issue an `addProgramEnrollment` api request.

#### Where should the reviewer start?

#### How should this be manually tested?

right now:

- go to the homepage
- run `window.openDialog()` in the JS console
- pick a program your user is not currently enrolled in
- go to the dashboard
- confirm that, after the dashboard loads, `localStorage.programId` is empty, and that your user has a new program enrollment.

#### Any background context you want to provide?

#### Screenshots (if appropriate)

#### What GIF best describes this PR or how it makes you feel?

noisecapella commented on static/js/signup_dialog.js:
```diff
@@ -0,0 +1,40 @@
+const dialogDiv = document.querySelector('#signup-dialog');
+
+const openDialog = () => store.dispatch(setSignupDialogVisibility(true));
+window.openDialog = openDialog;
```
Is this just for debugging?

aliceriot replied:
yes

aliceriot commented on static/js/util/test_utils.js:
```diff
@@ -1,7 +1,36 @@
+  });
+
+  const reset = () => {
+    [getItem, setItem, removeItem].forEach(func => {
```
Can you give an example of how I could make that work? I tried playing around with a sandbox and I couldn't get it to play nicely with the spy functions.

noisecapella replied:
At the top of `localStorageMock` you would put `const sandbox = sinon.sandbox.create()`. Then for example `getItem` would become `const getItem = sandbox.spy(...)`. I've never actually used it with spies, just stubs, but I feel like that should work.

aliceriot replied:
I tried the following in the nodejs repl:

```js
var sinon = require('sinon')

const sandbox = sinon.sandbox;

const spyFunc = x => x + 2

spyFunc(2) // 4
let spy = sandbox.spy(spyFunc)

spy.called // false

spy(2) // 4

spy.called // true

sandbox.restore()

spy.called // still true

sandbox.reset()

spy.called // still true
```

so it seems like the `restore` and `reset` methods on the sandbox don't clear and reset the spys and mocks created by the sandbox, it looks like it just clears the sandbox out so that new mocks may be added (`sandbox.fakes === []`).

noisecapella replied:
Oh that's interesting. I guess there's no way around this

aliceriot replied:
yeah I was kinda bummed, seemed like not exactly the way the sandbox should work...

aliceriot replied:
nvm! I forgot to called `sinon.sandbox.create()` above >_<

it actually does work, haha.

noisecapella commented on static/js/signup_dialog.js:
```diff
@@ -0,0 +1,40 @@
+logoLink.onClick = openDialog;
+
+if ( typeof SETTINGS.program === 'number' ) {
+  store.dispatch(setSignupProgram(SETTINGS.program));
```
Since the reducers have access to `SETTINGS` I feel like it would be cleaner to set the initial value there instead.

noisecapella commented on cms/models.py:
```diff
@@ -17,6 +17,15 @@
 from profiles.api import get_social_username
 from ui.views import get_bundle_url
 
+def program_info(programs):
```
Can we use the existing `ProgramSerializer` for this? It provides the same two fields. You should be able to use it like `ProgramSerializer().to_representation(program)`

aliceriot replied:
yeah that makes sense to me.

noisecapella commented on backends/pipeline_api.py:
```diff
@@ -72,6 +72,7 @@ def update_profile_from_edx(backend, user, response, is_new, *args, **kwargs):
         pass
     user_profile.gender = user_profile_edx.get('gender')
     user_profile.edx_mailing_address = user_profile_edx.get('mailing_address')
+    user_profile.agreed_to_terms_of_service = True
```
Currently the server side validation will allow you to switch this back to `False` as long as `filled_out` is not `True`. Since it doesn't seem like the user has a reason to set this anymore via PATCH, can we add it to the list of read only fields in `ProfileSerializer` and `ProfileLimitedSerializer`?

aliceriot replied:
yes that's a good idea

noisecapella commented on static/js/containers/App.js:
```diff
@@ -60,6 +64,7 @@ class App extends React.Component {
     this.fetchEnrollments();
     this.requireProfileFilledOut();
     this.requireCompleteProfile();
+    this.updateProgramEnrollments();
```
This should go in `componentDidMount` too

noisecapella commented:
It looks like localStorage is used entirely within React classes. What do you think about using redux-localstorage instead and putting the new program id in its own reducer?

noisecapella commented on static/js/containers/App.js:
```diff
@@ -91,6 +96,30 @@ class App extends React.Component {
+          if ( e.errorStatusCode === 404 ) {
+            window.localStorage.removeItem("programId");
+          } else {
+            window.localStorage.setItem("programIdFailed", programId);
```
I'm not sure if it makes sense to store this if we're not going to use it again. If we want to alert the user to a failure we should do a `console.error` instead or something more visible (cc @roberthouse54)

aliceriot replied:
Yeah I agree, I think we need more clarity on what sort of error handling behavior to use.

aliceriot commented:
I was thinking about that, I think it would make sense to pull out the code we need into a separate reducer both for simplicity and also to keep the bundle size down.

aliceriot commented:
@noisecapella I just pushed the changes to use `redux-localstorage` on the home page. I also moved the signup-dialog reducers and stuff into separate files (so we don't need to include the rest of the redux code in the bundle) and refactored `configureStore.js` to support the changes.

noisecapella commented:
Ok, let me know when it's ready for review

aliceriot commented:
will do!

aliceriot commented:
@noisecapella do you have any thoughts about testing the `SignupDialog` component? It's a container component (connected to a store) but it's not rendered inside of react-router. I thought about just making a store and creating some test infrastructure using `<Provider>` for it.

noisecapella commented:
Makes sense to me

aliceriot commented:
word, I'll give that a go and ping you when it's working.

aliceriot commented:
@noisecapella this is ready for review again

noisecapella commented:
This doesn't currently pick the newly added program as the current program, though that's coming as part of #1000 so no need to do anything here

noisecapella commented on static/js/actions/signup_dialog.js:
```diff
@@ -0,0 +1,9 @@
+// actions for the signup dialog on the home and program pages
+import { createAction } from 'redux-actions';
+
+export const SET_DIALOG_VISIBILITY = 'SET_SIGNUP_DIALOG_VISIBILITY';
```
Can you change the text for this and `SET_PROGRAM` to match the variable name?

aliceriot replied:
oops, that was a half-completed small refactor

noisecapella commented on static/js/containers/App.js:
```diff
@@ -102,6 +107,32 @@ class App extends React.Component {
+          e => {
+            window.localStorage.removeItem("signupDialogRedux");
+            if ( e.errorStatusCode !== 404 ) {
+              window.localStorage.setItem("programIdFailed", programId);
```
Can you change this to a `console.error` unless we have a use case for leaving it in localStorage?

aliceriot replied:
oh yeah, forgot to change that.

noisecapella commented on static/js/containers/SignupDialog.js:
```diff
@@ -0,0 +1,104 @@
+      <div className="terms-of-service-text">
+        By clicking "Continue with edX" I certify that I agree with
+        <a href="/terms_of_service" target="_blank">
+          {" MIT MicroMasters Terms of Service. "}
```
Just nit picking here but the leading space will show up in the link here and below

aliceriot replied:
Yeah that's intentional, so that there's a space between `with` from the line above and `MIT`.

noisecapella commented on static/js/containers/SignupDialog.js:
```diff
@@ -0,0 +1,104 @@
+});
+
+const mapDispatchToProps = dispatch => {
+  const setProgramHelper = (e,k,v) => {
```
Can you change `e,k,v` to `event, key, value` just for clarity?

noisecapella commented on static/js/containers/SignupDialog.js:
```diff
@@ -0,0 +1,104 @@
+  program: state.program,
+});
+
+const mapDispatchToProps = dispatch => {
```
This is a useful function, I keep forgetting `connect` takes a second argument

noisecapella commented on static/js/reducers/signup_dialog_test.js:
```diff
@@ -0,0 +1,52 @@
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    store = configureTestStore(signupDialog);
+    dispatchThen = store.createDispatchThen(state => state);
```
I think `state => state` is the default argument so you can leave it out here

noisecapella commented on static/js/store/configureStore.js:
```diff
@@ -4,29 +4,52 @@ import { compose, createStore, applyMiddleware } from 'redux';
+  compose(filter([path]))(adapter(window.localStorage))
+);
+
+const createPersistantStore = persistance => compose(
```
`persistant` => `persistent`, same for `persistance`

aliceriot replied:
oops, spelling :D

noisecapella commented on static/js/store/configureStore.js:
```diff
@@ -4,29 +4,52 @@ import { compose, createStore, applyMiddleware } from 'redux';
+  devTools()
+)(createStore);
+
+const createPersistantTestStore = persistance => compose(
```
Also here

noisecapella commented:
Looks good overall. I think if we can make use of the same program signup reducer in both bundles we won't need to mock localStorage or parse its JSON anymore

aliceriot commented:
Cool. I like the idea of refactoring to use the same reducer in both bundles, but I want to keep the localStorage mocking (and use it in unit tests) to test the behavior of the code in `App.js` to update the API if there's a saved program id.

aliceriot commented:
@noisecapella I made those changes, let me know what you think!

noisecapella commented:
The visibility of the dialog is stored in localStorage which is a little quirky to me, but feel free to leave it

noisecapella commented on static/js/containers/SignupDialog_test.js:
```diff
@@ -0,0 +1,81 @@
+      return { id: value, title: primaryText };
+    });
+
+    assert.deepEqual(renderedSelectOptions, SETTINGS.programs);
```
Should we sort the programs listed in the dialog?

aliceriot replied:
we could, alphabetically by the title?

noisecapella replied:
Sounds good to me, that's how I did it in the other PR

aliceriot replied:
word, I'll add that6

noisecapella commented:
I just merged #1000, can you rebase?

noisecapella commented on static/js/store/configureStore.js:
```diff
@@ -3,30 +3,59 @@
+  devTools()
+)(createStore);
+
+const createPersistantTestStore = persistence => compose(
```
Should be `createPersistentTestStore`

aliceriot replied:
lol idk why I have so much trouble with that word! haha

noisecapella commented:
Functionality works great

aliceriot commented:
@noisecapella comments should be addressed!

noisecapella commented:
:+1:
